content/blog/2025-06-06-a-plan-for-simd.md

```diff
@@ -0,0 +1,142 @@
+
+Another possibility is to evolve pulp in the direction we need.
+That’s still not out of the question, but the changes proposed are quite extensive, and this could be disruptive to the existing user base, particularly faer.
+One goal in publishing this plan is to gather feedback from the pulp community about what they’d like to see happen.
+
+## On RVV and SVE
+
+One topic I didn’t cover in my blog post is RVV and SVE, which are pretty marginal these days but will become more important.
+There are some pretty big challenges, and for the most part we’re blocked on Rust support for the intrinsics.
+
+Taking a longer view, I think there are two modes for “scalable vector” SIMD.
+One is to generate asm which operates correctly no matter the SIMD width.
+It’s fairly easy to see how to do this for map-like workloads, and I think it’s reasonable to consider this mostly a codegen problem for autovectorization, as opposed to a good use case for explicit SIMD.
+
+The other is to treat scalable vectors as an annoying instruction encoding for SIMD width not known at compile time, and with partial support for operating on fixed-width (128 bit) blocks.
+This is essentially the same concept as “[Fixed Width Vector Workloads on Variable Width Vectors](https://gist.github.com/zingaburga/805669eb891c820bd220418ee3f0d6bd#fixed-width-vector-workloads-on-variable-width-vectors)”.
+The entire concept is a bit frustrating because nearly all ARM chips are 128 bits anyway.
+
+See also the paper [SIMD Everywhere Optimization from ARM NEON to RISC-V Vector Extensions](https://arxiv.org/pdf/2309.16509) for an approach to making explicit SIMD code portable to RVV.
+
+## Considerations for WASM SIMD
+
+There’s a [Zulip thread](https://xi.zulipchat.com/#narrow/channel/197075-vello/topic/WASM.20SIMD) on this.
+One tricky bit is that WASM doesn’t have runtime feature detection, rather they expect feature detection to be done as part of the negotiation for deciding which WASM blob to serve.
+In some ways, this makes sense, as it avoids pretty much all of the difficulties of multiversioning (including potential binary size impact), but on the other hand it’s a pain, as it complicates build and deployment.
+
+## Very small SIMD
+
+In the embedded world, there are a variety of SIMD extensions for small widths.
+I believe the standard RVV extension (intended for application class chips) requires a width of at least 128 bits, but there is a Zve64 profile intended for embedded.
+Additionally, embedded ARM processors have a SIMD extension called “[Helium](https://en.wikichip.org/wiki/arm/helium)”, which is still 128 bit but with a smaller register count and more limited instructions.
+Given that Linebender is pushing more into `no_std`, it’s possible we’ll want to support these.
+However, I’m not able to find much evidence of Rust support for these.
+At this point, probably best to consider it an open question and potential future work.
+
+## Discussion
```

Comment:
Not just discussion ... but what are the next steps? What can people do? Should people try to use it?